### ISSUE TITLE
Add pagination

### ISSUE THREAD

**Issue by dfarr** [resonatehq/resonate#8]
APIs that require pagination
- SearchPromises
- ReadSubscriptions

**Comment by dfarr** [comment]
# Stable pagination

If we plan to support stable pagination we can rule out offset based pagination right away. Cursor based pagination could be based on a unique column that is known to only increase (or decrease) over time. At first glance, one of our timestamp columns seem like a good candidate.

Promises have two timestamps:
- created_on
- completed_on

However, using these columns would prove problematic for two reasons:
1. Both values are set to the tick time and therefore often not unique (a batch of promises are created/completed on each tick).
2. The tick time is not guaranteed to always increase.

The second point is especially tricky. While it is true that the tick time is guaranteed to only increase within a single process, what happens if my application restarts? What happens if I scale up my application and one of my instances clock is behind the other?

Instead, we can opt to add an auto incrementing column to our table and use this in our query, this is sometimes called keyset based pagination.
```sql
SELECT
  *
FROM
  promises
WHERE
  auto_incrementing_id < $1
ORDER BY
  auto_incrementing_id
LIMIT
  $2
```

# Non stable pagination

Easier to implement, possibly more performant. Can our client tolerate duplicate results?

**Comment by dtornow** [comment]
Requirements for pagination:

- a promise may be included twice
- a promise must *not* be omitted

**Comment by dfarr** [comment]
Closed by #71 and #66
